stl/inc/xhash

```diff
@@ -1915,6 +1915,166 @@ protected:
+                return {};
+            }
+        }
+
+        if
+            _CONSTEXPR_IF(_Traits::_Standard) {
+                // trim matching prefixes
+                while (*_First1 == *_First2) {
+                    // the right equal_range ends at the end of the bucket or on the first nonequal element
+                    bool _Right_range_end = _First2 == _Bucket_hi;
+                    ++_First2;
+                    if (!_Right_range_end) {
+                        _Right_range_end = _Right._Traitsobj(_Keyval, _Traits::_Kfn(*_First2));
+                    }
+
+                    // the left equal_range ends at the end of the container or on the first nonequal element
+                    ++_First1;
+                    const bool _Left_range_end =
+                        _First1 == _Unchecked_end() || _Traitsobj(_Keyval, _Traits::_Kfn(*_First1));
+
+                    if (_Left_range_end && _Right_range_end) {
+                        // the equal_ranges were completely equal
+                        return {true, _First1};
+                    }
+
+                    if (_Left_range_end || _Right_range_end) {
+                        // one equal_range is a prefix of the other; not equal
+                        return {};
+                    }
+                }
+
+                // found a mismatched element, find the end of the equal_ranges and dispatch to _Check_match_counts
+                auto _Last1 = _First1;
+                auto _Last2 = _First2;
+                for (;;) {
+                    bool _Right_range_end = _Last2 == _Bucket_hi;
+                    ++_Last2;
+                    if (!_Right_range_end) {
+                        _Right_range_end = _Right._Traitsobj(_Keyval, _Traits::_Kfn(*_Last2));
+                    }
+
+                    ++_Last1;
+                    const bool _Left_range_end =
+                        _Last1 == _Unchecked_end() || _Traitsobj(_Keyval, _Traits::_Kfn(*_Last1));
+
+                    if (_Left_range_end && _Right_range_end) {
+                        // equal_ranges had the same length, check for permutation
+                        return {_Check_match_counts(_First1, _Last1, _First2, _Last2, equal_to<>{}), _Last1};
+                    }
+
+                    if (_Left_range_end || _Right_range_end) {
+                        // different number of elements in the range, not a permutation
+                        return {};
+                    }
+                }
+            }
+        else {
+            // check the first elements for equivalence when !_Standard
+            if (_Right._Traitsobj(_Keyval, _Traits::_Kfn(*_First2))) {
+                return {};
+            }
+
+            // trim matching prefixes
+            const size_t _LHashval   = _Traitsobj(_Keyval);
+            const size_type _LBucket = _LHashval & _Mask;
+            const auto _LBucket_hi   = _Vec._Mypair._Myval2._Myfirst[(_LBucket << 1) + 1];

```

Comment:
If I had sufficient coffee this time this is the main difference between the two `if constexpr` branches.

I you would store the pointer after `_LBucket_hi` you could also write

```cpp
    ++_First1;
    const bool _Left_range_end =  
        _First1 == _LBucket_hi_next || _Traitsobj(_Keyval, _Traits::_Kfn(*_First1));
```

With that the only difference between the two loops (besides comments) is `_LBucket_hi_next ` vs `_Unchecked_end()`. Given that you use a temporary variable anyway you could assign `_Unchecked_end()` to `_LBucket_hi_next ` in the `_Traits::_Standard` case

Dunno if that is truly worth it

Reply:
That will do one extra pointer dereference per bucket (to find `_LBucket_hi_next`); but maybe we don't care about the perf of `!_Standard` (that is, `hash_foo` rather than `unordered_foo`)  ...

Reply:
Looks good, applied.